Add support for automatic building of Makefile projects

lenticularis39 opened the pull request:
This patch adds a `build` command in addition to `configure` and `generate`. This commands takes a `Makefile` project and automatically builds it into a DiffKemp snapshot using an integrated GCC wrapper.

viktormalik commented on diffkemp/building/cc_wrapper_template.py:
```diff
@@ -0,0 +1,213 @@
+    # Analyze and modify parameters for clang (phase 2)
+    clang_argv[0] = clang
+    if clang != llvm_link:
+        for arg in append:
+            clang_argv.append(arg)
```
Not sure if RPython allows this:
```suggestion
        clang_argv.extend(append)
```

viktormalik commented on diffkemp/building/cc_wrapper_template.py:
```diff
@@ -0,0 +1,213 @@
+        # Compile/link mode with object files detected
+        # Drop object files and revert to normal compile/link mode
+        clang = old_clang
+        new_clang_argv = []
+        for arg in clang_argv:
+            if arg.endswith('.ll'):
+                continue
+            new_clang_argv.append(arg)
+        clang_argv = new_clang_argv
```
```suggestion
        clang_argv = [a for a in clang_argv if a.endswith('.ll')]
```

viktormalik commented on diffkemp/building/cc_wrapper_template.py:
```diff
@@ -0,0 +1,213 @@
+        db.append(prefix + os.path.join(os.getcwd(), output_file))
+    elif not linking:
+        # Compiling to default output file
+        for arg in clang_argv:
+            if not arg.endswith('.c'):
+                continue
+            db.append('o:' + os.path.join(os.getcwd(),
+                                          arg.rsplit('.', 1)[0] + '.ll'))
```
If RPython allows `extend`:
```suggestion
        db.extend(['o:' + os.path.join(...) if not arg.endswith('.c')])
```

viktormalik commented on diffkemp/building/cc_wrapper_template.py:
```diff
@@ -0,0 +1,213 @@
+                continue
+            db_file.write(entry + '\n')
+
+    # Disassemble output when running llvm-link
```
Does this assume that `llvm-link` always produces assembled (`.bc`) files? Is that really true?

lenticularis39 replied:
Yes, unless the `-S` option is used, I could use that instead of disassembling with `llvm-dis`.

viktormalik commented on diffkemp/building/cc_wrapper_template.py:
```diff
@@ -0,0 +1,213 @@
+    try:
+        check_call('gcc', argv)
+    except CalledProcessError:
+        return 1
```
Same as above: a message would be helpful.

viktormalik commented on diffkemp/building/cc_wrapper_template.py:
```diff
@@ -0,0 +1,213 @@
+    llvm_dis = 'llvm-dis'
+
+    # Run GCC
+    argv[0] = 'gcc'
```
Why GCC? What if the original command runs clang? Don't we want just to run the original here?

viktormalik reviewed (commented):
This is a very useful improvement, many thanks for it!

In general, I'm happy with the introduction of the `build` subcommand, but I'd like to think a bit more about possible re-organization of the existing commands. Currently, we have 3 builders:
- kernel builder, which takes no options
- single LLVM IR builder, which takes one param (path to the file)
- the new "wrapper builder" (can we find a better name, perhaps?), which takes a lot of params

In future, this could be possibly extended (for instance, we could support projects already compiled into LLVM IR - i.e. something like what the new builder does without the actual build).

The question now is, if we should have all of these under one command (`build`?), under 2 commands (`generate` and `build`?), or have a separate command for each.

lenticularis39 commented on diffkemp/building/cc_wrapper_gen.py:
```diff
@@ -0,0 +1,23 @@
+
+def generate_cc_wrapper(rpython='rpython', builddir='build'):
+    wrapper_template = 'cc_wrapper_template'
+    compiled_wrapper_filename = os.path.join(os.getcwd(), builddir,
+                                             wrapper_template + '-c')
```
It's the suffix generated by default by RPython, it can be changed with `--output`, I just simply decided to go with the default here.

viktormalik replied:
ok, makes sense

lenticularis39 commented on diffkemp/building/cc_wrapper_gen.py:
```diff
@@ -0,0 +1,23 @@
+
+def generate_cc_wrapper(rpython='rpython', builddir='build'):
+    wrapper_template = 'cc_wrapper_template'
+    compiled_wrapper_filename = os.path.join(os.getcwd(), builddir,
+                                             wrapper_template + '-c')
```
I didn't like that it would recompile the wrapper each time `/tmp` is cleared, however a proper permanent solution would require moving the generation of the wrapper out of generate phase into DiffKemp's build system.

Now thinking about it again, some systems have `/tmp` mounted with `noexec`, which is a bigger issue. I'll move the wrapper to be built and installed by CMake together with SimpLL, that should work.

lenticularis39 commented on diffkemp/building/cc_wrapper_template.py:
```diff
@@ -0,0 +1,213 @@
+            output_file = arg
+        elif is_object_file and linking:
+            # Input to linking phase: change suffix to .ll
+            arg = arg.rsplit('.', 1)[0] + '.ll'
```
Yes, that's just how the logic is composed, differentiating between input files and the output file.

lenticularis39 commented on diffkemp/building/cc_wrapper_template.py:
```diff
@@ -0,0 +1,213 @@
+
+    # Record file in database
+    if output_file is not None:
+        prefix = 'o:' if clang != llvm_link else 'f:'
```
The wrapper only sets `clang` equal to `llvm_link` when object files are being linked, not when all arguments are C sources (both have `linking` set to `True`, since there's no `-c` argument present in either case), so these are not equivalent. 

The wrapper logic readability should be improved though, the naming and the branches are quite confusing to read.

lenticularis39 commented:
Rebased on top of #241. Also implemented most of the review notes, excluding detecting original compiler (that is a little bit tricky). README is not updated yet, I'll do that in another commit.

viktormalik commented on diffkemp/building/cc_wrapper_gen.py:
```diff
@@ -0,0 +1,23 @@
+import os
+
+
+def generate_cc_wrapper(rpython='rpython', builddir='build'):
```
We're using double quotes everywhere, could you follow that?

lenticularis39 replied:
Actually `grep -r "'"` shows that we are not always using double quotes, a few lines in `diffkemp/cli.py` and `diffkemp/syndiff/function_syntax_diff.py` use single quotes, that is probably not intentional though. My use of single quotes wasn't intentional either, I guess I was influenced by an RPython code example I used for `cc_wrapper_template.py`.

Either way I'll change the occurences of single quotes in this patch to double quotes, the rest can be dealt with later if the affected code is changed.

viktormalik replied:
Yes, agreed.

lenticularis39 commented on diffkemp/diffkemp.py:
```diff
@@ -26,6 +32,90 @@ def run_from_cli():
+    if args.functions_list is None:
+        args.functions_list = os.path.join(args.source_dir, 'function_list')
+    with open(args.functions_list, "r") as fun_list_file:
+        function_list = [line.rstrip() for line in fun_list_file.readlines()]
```
Good catch, perhaps I was too quick with the rebasing to new-generate-cli branch :D

lenticularis39 replied:
Also I noticed that the argument probably should be consistently named `symbol_list`, not `function_list` among subcommands, I'll change that to get it consistent with build subcommand at least.

lenticularis39 commented on diffkemp/diffkemp.py:
```diff
@@ -26,6 +32,90 @@ def run_from_cli():
+            if configure_cmd and make_cc_setting not in configure_cmd:
+                configure_cmd = configure_cmd.split('$ ', 1)[1]
+                configure_cmd += ' ' + make_cc_setting
+                check_call('find "{}" -name config.cache | xargs rm -f'
+                           .format(args.source_dir), shell=True)
```
Not much, actually it would be about the same code length. Only con is Python syntax with loops/filter/map being slightly more difficult to read, but that's not enough reason to call external programs instead.

viktormalik replied:
I agree, let's use the Python way where possible (and make it as readable as we can).

lenticularis39 commented on diffkemp/diffkemp.py:
```diff
@@ -26,6 +32,90 @@ def run_from_cli():
+                if '$' in line:
+                    configure_cmd = line.strip()
+                    break
+            if configure_cmd and make_cc_setting not in configure_cmd:
+                configure_cmd = configure_cmd.split('$ ', 1)[1]
+                configure_cmd += ' ' + make_cc_setting
```
Definitely, I also had to think about it for a while to remember what it does :D

lenticularis39 commented on diffkemp/cli.py:
```diff
@@ -12,6 +12,38 @@ def make_argument_parser():
+                          action='append')
+    build_ap.add_argument("--reconfigure", help="reconfigure autotools-based\
+                          project with CC=<wrapper>", action="store_true")
+    build_ap.add_argument("--without-rpython",
```
I was thinking about autotools configure, which uses `--without-...` to disable features while building, but thinking about it again, this is a different case and it could be confusing, `--no-rpython` is better here.

lenticularis39 commented on diffkemp/cli.py:
```diff
@@ -12,6 +12,38 @@ def make_argument_parser():
+                          help="list of functions to compare")
+    build_ap.add_argument("--make", help="make tool used to be used for build",
+                          default="make")
+    build_ap.add_argument("--makefile", help="filename of Makefile to be used\
+                          for build")
```
It doesn't have `default=Makefile`, because different versions of Make have their own rules for default makefile when they aren't explicitly given one (e.g. https://www.gnu.org/software/make/manual/html_node/Makefile-Names.html).

As of the default names, they are specific for Make, because generate now only supports Make. I'll change the names to generic ones though to avoid an interface change in the future when support for other tools is added (which it probably will, since Make is slowly becoming kind of obsolete).

viktormalik replied:
> It doesn't have `default=Makefile`, because different versions of Make have their own rules for default makefile when they aren't explicitly given one (e.g. https://www.gnu.org/software/make/manual/html_node/Makefile-Names.html).

Still, `Makefile` is "kind of" standard, so I'm slightly in favor of making it default.

> As of the default names, they are specific for Make, because generate now only supports Make. I'll change the names to generic ones though to avoid an interface change in the future when support for other tools is added (which it probably will, since Make is slowly becoming kind of obsolete).

Agreed, let's make them more generic (though I still wouldn't mind having `Makefile` as the default value).

lenticularis39 replied:
> 
> Still, `Makefile` is "kind of" standard, so I'm slightly in favor of making it default.
> 
IMO that would cause unexpected behavior in some cases, e.g. if `GNUmakefile` behaves differently than `Makefile` (either intentionally or unintentionally), `diffkemp build` with default settings would behave differently than `make` (provided `make` is GNU Make).
> Agreed, let's make them more generic (though I still wouldn't mind having `Makefile` as the default value).
With generic UI, it would have to be changed in the future anyway, because e.g. Ninja's defaut build file is `build.ninja` etc.

viktormalik replied:
Ok, let's leave it without the default, then.

lenticularis39 commented:
In addition to implementing the review comments, I also changed several things about the compiler wrapper:
- There is now only one file that is both called from make/built by RPython and included by diffkemp.py.
- Wrapper build with RPython does not happen in runtime, but is included into CMake instead; this makes it possible for the binary to be installed
- There are new rules for the wrapper, see `get_cc_wrapper_path` function
- `--no-rpython` was renamed to `--no-native-cc-wrapper`, since build subcommand no longer calls RPython

lenticularis39 commented:
There is one more issue I noticed now: pip won't install `cc_wrapper.py` as executable, which is understandable, but breaks the wrapper in case the native one isn't present and diffkemp is used installed by pip.

lenticularis39 commented:
> There is one more issue I noticed now: pip won't install `cc_wrapper.py` as executable, which is understandable, but breaks the wrapper in case the native one isn't present and diffkemp is used installed by pip.

Fixed, I set CMake to install the Python wrapper as `diffkemp-cc-wrapper.py` alongside the native `diffkemp-cc-wrapper` and updated `cc_wrapper.py` choice priority.

viktormalik commented on diffkemp/diffkemp.py:
```diff
@@ -18,222 +22,266 @@
+                configure_cmd += " " + make_cc_setting
+                # Delete all config.cache files
+                for root, _, files in os.walk(args.source_dir):
+                    for filename in files:
+                        if filename != "config.cache":
+                            continue
+                        try:
+                            os.remove(os.path.join(root, filename))
+                        except PermissionError:
+                            pass
```
Maybe more clear like this?
```suggestion
                    if "config.cache" in files:
                        try:
                            os.remove(os.path.join(root, "config.cache"))
                        except PermissionError:
                            pass
```

lenticularis39 replied:
Actually I should have simply iterated on directories and done something like
```python
try:
    os.remove(os.path.join(dirname, "config.cache")
except (FileNotFoundError, PermissionError):
    pass
```
I got confused by somehow automatically trying to replicate exactly what the shell one-liner did, which is less efficient on modern filesystems, since `os.remove` just looks up the file in a B-tree or a similar structure once, no need to go through all files in all directories in the tree.

viktormalik commented on diffkemp/llvm_ir/wrapper_build_finder.py:
```diff
@@ -0,0 +1,68 @@
+        # present in the o: files
+        llvm_files = []
+        with open(self.wrapper_db_filename, 'r') as db_file:
+            for line in db_file.readlines():
+                line = line.rstrip()
+                if not line.startswith('o:'):
+                    continue
+                line = line.split(':')[1]
+                llvm_files.append(line)
```
This feels more readable:
```suggestion
            llvm_files = [line.split(':')[1].rstrip()
                          for line in db_file.readlines()
                          if line.startswith(".o")]
```

viktormalik reviewed (commented):
Just two loop simplification suggestions, otherwise LGTM.
